Translation updater: Add support for function call without parantheses

srifqi opened the pull request:
**Goal of the PR**
This PR adds support for Lua function call without parentheses (single parameter function call).

**How does the PR work?**
This PR adds two patterns for Lua function call without parentheses, both for `"single/double-quote"` and for `[[multiline]]`.

**Does it resolve any reported issue?**
This PR tries to fix #14494.

**If not a bug fix, why is this PR needed? What usecases does it solve?**
See the linked issue and https://www.lua.org/pil/5.html for informations about single argument call.

## To do

This PR is Ready for Review.

## How to test

1. Create a mod with some code using `S` (and its variants) and without parentheses.
2. Run the script.
3. Check that the string exists.

```lua
S("Double-quote with parentheses")
S"Double-quote without parentheses"
S('Single-quote with parentheses')
S'Single-quote without parentheses'
S([[Multiline with parentheses]])
S[[Multiline without parentheses]]
S
[[Multiline
without
parentheses]]
```

srifqi commented:
**Note:**
The string order in the translation files depends on the way the strings are written:
1. Single/double-quote with parentheses
2. Multiline with parentheses
3. Single/double-quote **without** parentheses
4. Multiline **without** parentheses

I haven't had a success in combining all of those so that the order matches the occurrences in the file.

SmallJoker commented:
Modified devtest/chest/chest.lua
```Lua
local S = minetest.get_translator("chest")

minetest.register_node("chest:chest", {
	description = S"Chest" .. "\n" ..
		"32 inventory slots",
```

results in
```
# textdomain: chest
Chest@n32 inventory slots=
```

However, only "Chest" should be translated.

srifqi commented:
I moved matching strings for single parameter call before strings concatenations.

This results in string order change:

1. Single/double-quote **without** parentheses
2. Multiline **without** parentheses
3. Single/double-quote with parentheses
4. Multiline with parentheses

I also rebased this PR.

-------

Ordering strings based on their occurrences in the file will be harder since there are text changes due to string concatenations **after** matching function call without parentheses and **before** matching function call with parentheses.

SmallJoker reviewed (approved):
Works

appgurueu commented on util/mod_translation_updater.py:
```diff
@@ -133,6 +133,19 @@ def main():
+# Handles like pattern_lua_quoted, but for single parameter (without parantheses)
+# See https://www.lua.org/pil/5.html for informations about single argument call
+pattern_lua_quoted_single = re.compile(
+	r'(?:^|[\.=,{\(\s])' # Same as for pattern_lua_quoted
```
If you have these "same" parts, why not just extract them into variables with useful names?

srifqi replied:
> If you have these "same" parts, why not just extract them into variables with useful names?

Something like this?

<details>
<summary>Python code</summary>

```python
# List of pattern components:
# Look for beginning of file or anything that isn't a function identifier
pattern_lua_begin_non_function = r'(?:^|[\.=,{\(\s])'
# Matches S, FS, NS, or NFS function call
pattern_lua_NFS_function = r'N?F?S\s*'
# Start of argument: open parentheses and space (optional)
pattern_lua_open_parentheses = r'\(\s*'
# Quoted string
# Group 2 will be the string, group 1 and group 3 will be the delimiters (" or ')
# See https://stackoverflow.com/questions/46967465/regex-match-text-in-either-single-or-double-quote
pattern_lua_quoted_string = r'(["\'])((?:\\\1|(?:(?!\1)).)*)(\1)'
# Double square bracket string (multiline)
pattern_lua_square_bracket_string = r'\[\[(.*?)\]\]'
# End of argument or function call: space, comma, or close parentheses
pattern_lua_close_parentheses = r'[\s,\)]'

# Handles the " ... " or ' ... ' string delimiters
pattern_lua_quoted = re.compile(
	pattern_lua_begin_non_function +
	pattern_lua_NFS_function +
	pattern_lua_open_parentheses +
	pattern_lua_quoted_string +
	pattern_lua_close_parentheses,
	re.DOTALL)
# Handles the [[ ... ]] string delimiters
pattern_lua_bracketed = re.compile(
	pattern_lua_begin_non_function +
	pattern_lua_NFS_function +
	pattern_lua_open_parentheses +
	pattern_lua_square_bracket_string +
	pattern_lua_close_parentheses,
	re.DOTALL)
# Handles like pattern_lua_quoted, but for single parameter (without parentheses)
# See https://www.lua.org/pil/5.html for informations about single argument call
pattern_lua_quoted_single = re.compile(
	pattern_lua_begin_non_function +
	pattern_lua_NFS_function +
	pattern_lua_quoted_string,
	re.DOTALL)
# Same as pattern_lua_quoted_single, but for [[ ... ]] string delimiters
pattern_lua_bracketed_single = re.compile(
	pattern_lua_begin_non_function +
	pattern_lua_NFS_function +
	pattern_lua_square_bracket_string,
	re.DOTALL)
```
</details>

> Also, I see no reason why this can't be a single regex. Everything is common except for the "body", so you could just do ... common stuff ... (quoted string|bracketed string), right?

(See my other reply.)

> (Furthermore, I don't see why this wouldn't be able to share a pattern with parenthesized function calls even.)

I have not checked that.

srifqi replied:
Based on [the discussion over #minetest-dev](https://irc.minetest.net/minetest-dev/2024-05-12#i_6170441) with @appgurueu, this is the revised code:

<details>
<summary>Python code</summary>

```python
# Compile pattern for matching lua function call
def compile_func_call_pattern(argument_pattern):
	return re.compile(
		# Look for beginning of file or anything that isn't a function identifier
		r'(?:^|[\.=,{\(\s])' +
		# Matches S, FS, NS, or NFS function call
		r'N?F?S\s*' +
		# The pattern to match argument
		argument_pattern,
		re.DOTALL)

# Add parentheses around a pattern
def parenthesize_pattern(pattern):
	return (
		# Start of argument: open parentheses and space (optional)
		r'\(\s*' +
		# The pattern to be parenthesized
		pattern +
		# End of argument or function call: space, comma, or close parentheses
		r'[\s,\)]')

# Quoted string
# Group 2 will be the string, group 1 and group 3 will be the delimiters (" or ')
# See https://stackoverflow.com/questions/46967465/regex-match-text-in-either-single-or-double-quote
pattern_lua_quoted_string = r'(["\'])((?:\\\1|(?:(?!\1)).)*)(\1)'

# Double square bracket string (multiline)
pattern_lua_square_bracket_string = r'\[\[(.*?)\]\]'

# Handles the " ... " or ' ... ' string delimiters
pattern_lua_quoted = compile_func_call_pattern(parenthesize_pattern(pattern_lua_quoted_string))

# Handles the [[ ... ]] string delimiters
pattern_lua_bracketed = compile_func_call_pattern(parenthesize_pattern(pattern_lua_square_bracket_string))

# Handles like pattern_lua_quoted, but for single parameter (without parentheses)
# See https://www.lua.org/pil/5.html for informations about single argument call
pattern_lua_quoted_single = compile_func_call_pattern(pattern_lua_quoted_string)

# Same as pattern_lua_quoted_single, but for [[ ... ]] string delimiters
pattern_lua_bracketed_single = compile_func_call_pattern(pattern_lua_square_bracket_string)
```
</details>

appgurueu commented on util/mod_translation_updater.py:
```diff
@@ -272,9 +285,17 @@ def read_lua_file_strings(lua_file):
+		strings = []
+
+		for s in pattern_lua_quoted_single.findall(text):
+			strings.append(s[1])
```
You can avoid the odd `s[1]` if you don't use a backreference and capture, but instead construct the regex as `'((?:\\.|[^\\']))*'|"((?:\\.|[^\\"]))*"`.

srifqi replied:
That would result in even more capture groups (and would need checking each group's existence).

```python
import re
pattern = re.compile(r'(a)|(b)|(c)', re.DOTALL)
pattern.findall('abc')
# [('a', '', ''), ('', 'b', ''), ('', '', 'c')]
```

... or maybe we can just concatenate all since it is always one or the other.

```python
[''.join(x) for x in pattern.findall('abc')]
# ['a', 'b', 'c']
```

appgurueu replied:
Why write `(a)|(b)|(c)` instead of `((?:a)|(?:b)|(?:c))` (where `(?:x)` can be just `x` if precedency allows it)

srifqi replied:
I was thinking about using it for `"..."`, `'...'`, and `[[...]]` which we also want to remove the delimiters.

```python
import re

pattern = re.compile(r'"((?:\\.|[^\"])*)"|\'((?:\\.|[^\\\'])*)\'|\[\[(.*?)\]\]')
sample_string = '"ABC\'DEF\\"", \'GHI"JKL\\\'\', [[MNOPQR]]'

pattern.findall(sample_string)
# [('ABC\'DEF\\"', '', ''), ('', 'GHI"JKL\\\'', ''), ('', '', 'MNOPQR')]

[''.join(x) for x in pattern.findall(sample_string)]
# ['ABC\'DEF\\"', 'GHI"JKL\\\'', 'MNOPQR']
```

The alternative might be trimming delimiters from the captured text.

appgurueu commented:
(feel free to go ahead and merge this now)
